ports/stm32f4/common-hal/busio/UART.c

```diff
@@ -36,45 +36,455 @@
+            errflag = HAL_UART_Receive_IT(&self->handle, &self->rx_char, 1);
+        }
+        // Allow user to break out of a timeout with a KeyboardInterrupt.
+        if ( mp_hal_is_interrupted() ) {
+            return 0;
+        }
+    }
+
+    // Halt reception
+    HAL_NVIC_DisableIRQ(self->irq);
+    // copy received data
+    rx_bytes = ringbuf_count(&self->rbuf);
+    //Used for debugging
+    //mp_printf(&mp_plat_print, "Read: count:%d, buffer location%p, er:%d\n", rx_bytes, &self->rbuf,errflag);
+    rx_bytes = MIN(rx_bytes, len);
+    for ( uint16_t i = 0; i < rx_bytes; i++ ) {
+        data[i] = ringbuf_get(&self->rbuf);
+    }
+    HAL_NVIC_EnableIRQ(self->irq);
+
+    if (rx_bytes == 0) {
+        *errcode = EAGAIN;
+        return MP_STREAM_ERROR;
+    }
+    return rx_bytes; 
 }
 
 // Write characters.
 size_t common_hal_busio_uart_write(busio_uart_obj_t *self, const uint8_t *data, size_t len, int *errcode) {
-    return 0;
+    if (self->tx == NULL) {
+        mp_raise_ValueError(translate("No TX pin"));
+    }
+
+    HAL_NVIC_DisableIRQ(self->irq);
+
+    if (HAL_UART_Transmit(&self->handle, (uint8_t *)data, len, self->timeout_ms) != HAL_OK) {
+        mp_raise_ValueError(translate("UART write error"));
+    }
+
+    HAL_UART_Receive_IT(&self->handle, &self->rx_char, 1);
+    HAL_NVIC_SetPriority(self->irq, UART_IRQPRI, UART_IRQSUB_PRI);
+    HAL_NVIC_EnableIRQ(self->irq);
+
+    return len;
+}
+
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *handle)
+{
+    for(int i=0; i<7; i++) {
+        //get context pointer and cast it as struct pointer
+        busio_uart_obj_t * context = (busio_uart_obj_t *)MP_STATE_PORT(cpy_uart_obj_all)[i];
+        if(handle == &context->handle) {
+            //check if transaction is ongoing
+            if((HAL_UART_GetState(handle) & HAL_UART_STATE_BUSY_RX) == HAL_UART_STATE_BUSY_RX) {
+                return;
+            }
+            ringbuf_put_n(&context->rbuf, &context->rx_char, 1);
+            errflag = HAL_UART_Receive_IT(handle, &context->rx_char, 1);
+            
+            return;
+        }
+    }
+}
+
+void HAL_UART_ErrorCallback(UART_HandleTypeDef *UartHandle)
+{
+    if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_PE) != RESET) {
+        __HAL_UART_CLEAR_PEFLAG(UartHandle);
+    } else if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_FE) != RESET) {
+        __HAL_UART_CLEAR_FEFLAG(UartHandle);
+    } else if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_NE) != RESET) {
+        __HAL_UART_CLEAR_NEFLAG(UartHandle);
+    } else if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_ORE) != RESET) {
+        __HAL_UART_CLEAR_OREFLAG(UartHandle);
+    }
+    //restart serial read after an error
+    for(int i=0; i<7; i++) {
+        busio_uart_obj_t * context = (busio_uart_obj_t *)MP_STATE_PORT(cpy_uart_obj_all)[i];
+        if(UartHandle == &context->handle) {
+            HAL_UART_Receive_IT(UartHandle, &context->rx_char, 1);
+            return;
+        }
+    }
+
```

Comment:
This is kind of crazy because it starts and stops reception every read.

What about intercepting the RX related interrupts (https://github.com/adafruit/stm32f4/blob/89a356f0b44883d55c45a5d22940777a44a2b881/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c#L2011) and handling them yourself? More below in call_hal_irq.

Reply:
Hi @tannewt I considered doing the RX interception, but opted to base it off stm32duino instead, to avoid exposing all the error management and parity/wordlength/buffer code. I could probably make it more elegant/speedy with a register approach, but perhaps that could be pushed further down the line with an issue? 

I'm not sure exactly what you mean by stop and start - most of the time `HAL_UART_Receive_IT(handle, &context->rx_char, 1)` will transition into the next read the same way it does in the NRF code and identically to stm32duino. Can you elaborate on your concern?

Reply:
My concern was that the HAL handler [stops the reception](https://github.com/adafruit/stm32f4/blob/89a356f0b44883d55c45a5d22940777a44a2b881/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c#L3004-L3022) before calling the RxComplete callback and leaves a gap before it is re-enabled by the next call to Receive_IT is called. If a byte is received in this gap then it will be missed. This probably isn't a problem for UART data because the gap is quite short. However, there is a gap that isn't needed.